java-checks/src/main/java/org/sonar/java/checks/SpecializedFunctionalInterfacesCheck.java

```diff
@@ -0,0 +1,230 @@
+    ParameterTypeNameAndTreeType firstArgument = new ParameterTypeNameAndTreeType(args.get(0));
+    ParameterTypeNameAndTreeType secondArgument = new ParameterTypeNameAndTreeType(args.get(1));
+    String reportString = null;
+    if (firstArgument.paramType.equals(secondArgument.paramType)) {
+      reportString = String.format("UnaryOperator<%s>", firstArgument.paramType.name());
+    } else if (firstArgument.paramTypeName != null && secondArgument.paramTypeName != null) {
+      reportString = String.format("%sTo%sFunction", firstArgument.paramTypeName, secondArgument.paramTypeName);
+    } else if (firstArgument.paramTypeName == null && secondArgument.paramTypeName != null) {
+      reportString = String.format("To%sFunction<%s>", secondArgument.paramTypeName, firstArgument.paramType.name());
+    } else if (firstArgument.paramTypeName != null) {
+      reportString = String.format("%sFunction<%s>", firstArgument.paramTypeName, secondArgument.paramType.name());
+    }
+    return Optional.ofNullable(reportString);
+  }
+
+  private static Optional<String> handleBiFunctionInterface(TypeArguments args) {
+    ParameterTypeNameAndTreeType firstArgument = new ParameterTypeNameAndTreeType(args.get(0));
+    ParameterTypeNameAndTreeType secondArgument = new ParameterTypeNameAndTreeType(args.get(1));
+    ParameterTypeNameAndTreeType thirdArgument = new ParameterTypeNameAndTreeType(args.get(2));
+    if (firstArgument.paramType.equals(secondArgument.paramType) && firstArgument.paramType.equals(thirdArgument.paramType)) {
+      return Optional.of("BinaryOperator<" + firstArgument.paramType.name() + ">");
+    }
+    return Optional.empty();
+  }
+
+  private static Optional<String> handleBiConsumerInterface(TypeArguments args) {
+    ParameterTypeNameAndTreeType firstArgument = new ParameterTypeNameAndTreeType(args.get(0));
+    ParameterTypeNameAndTreeType secondArgument = new ParameterTypeNameAndTreeType(args.get(1));
+    if (secondArgument.paramTypeName != null) {
+      return Optional.ofNullable(String.format("Obj%sConsumer<%s>", secondArgument.paramTypeName, firstArgument.paramType));
+    }
+    return Optional.empty();
+  }
+
+  private static Optional<TypeArguments> interfaceParameters(TypeTree interfaceType) {
+    if (interfaceType.is(Tree.Kind.PARAMETERIZED_TYPE)) {
+      return Optional.of(((ParameterizedTypeTree) interfaceType).typeArguments());
+    }
+    return Optional.empty();
+  }
+
+  private static class InterfaceTreeAndStringPairReport {
+    final String reportString;
+    final TypeTree classInterface;
+
+    public InterfaceTreeAndStringPairReport(String report, TypeTree interf) {
+      reportString = report;
+      classInterface = interf;
+    }
+  }
+
+  private static class ParameterTypeNameAndTreeType {
+
+    Type paramType;
+
+    @Nullable
+    String paramTypeName;
+
+    public ParameterTypeNameAndTreeType(Tree tree) {
+      switch (tree.kind()) {
```

Comment:
the default is not covered. Did you tried? (it is) If it's not possible, then throw an IllegalStateException in such case, with an adequate message (look in other checks).